### PR TITLE
Catch UnicodeEncodeErrors in _add_flaky_report()

### LINKED ISSUE #47: UnicodeEncodeError when raising an exception with non-ascii characters
The Python 2.6 tests will fail with a UnicodeEncodeError if an exception is raised with non-ascii characters in the message.

For example, I updated the test_non_flaky_thing() test in test_example.py to include the non-ascii message in the exception:

```python
@genty
class ExampleFlakyTestsWithUnicodeTestNames(ExampleFlakyTests):
    @genty_dataset('ascii name', 'ńőń ȁŝćȉȉ ŝƭȕƒƒ')
    def test_non_flaky_thing(self, message):
        self._threshold += 1
        if self._threshold < 1:
            raise Exception(
                "Threshold is not high enough: {0} vs {1} for '{2}'.".format(
                    self._threshold, 1, message),
            )
```

This caused a UnicodeEncodeError when running the Python 2.6 tests:

```
py26 runtests: PYTHONHASHSEED='2117695302'
py26 runtests: commands[0] | nosetests --with-flaky --exclude=pytest|test_options_example
....................................
===Flaky Test Report===

Traceback (most recent call last):
  File ".tox/py26/bin/nosetests", line 11, in <module>
    sys.exit(run_exit())
  File "/Users/awee/open_source/flaky_fork/flaky/.tox/py26/lib/python2.6/site-packages/nose/core.py", line 121, in __init__
    **extra_args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.6/lib/python2.6/unittest.py", line 817, in __init__
    self.runTests()
  File "/Users/awee/open_source/flaky_fork/flaky/.tox/py26/lib/python2.6/site-packages/nose/core.py", line 207, in runTests
    result = self.testRunner.run(self.test)
  File "/Users/awee/open_source/flaky_fork/flaky/.tox/py26/lib/python2.6/site-packages/nose/core.py", line 66, in run
    result.printErrors()
  File "/Users/awee/open_source/flaky_fork/flaky/.tox/py26/lib/python2.6/site-packages/nose/result.py", line 110, in printErrors
    self.config.plugins.report(self.stream)
  File "/Users/awee/open_source/flaky_fork/flaky/.tox/py26/lib/python2.6/site-packages/nose/plugins/manager.py", line 99, in __call__
    return self.call(*arg, **kw)
  File "/Users/awee/open_source/flaky_fork/flaky/.tox/py26/lib/python2.6/site-packages/nose/plugins/manager.py", line 167, in simple
    result = meth(*arg, **kw)
  File "/Users/awee/open_source/flaky_fork/flaky/flaky/flaky_nose_plugin.py", line 182, in report
    self._add_flaky_report(stream)
  File "/Users/awee/open_source/flaky_fork/flaky/flaky/_flaky_plugin.py", line 236, in _add_flaky_report
    stream.write(self._stream.getvalue())
UnicodeEncodeError: 'ascii' codec can't encode characters in position 760-762: ordinal not in range(128)
ERROR: InvocationError: '/Users/awee/open_source/flaky_fork/flaky/.tox/py26/bin/nosetests --with-flaky --exclude=pytest|test_options_example'
__________________________________________________________________________________ summary __________________________________________________________________________________
ERROR:   py26: commands failed
```

### PR DESCRIPTION
**Pull request by anthonywee**
Fixes #47.
Catch a UncodeEncodeError when writing to the stream, and encode using
utf-8 if needed.

### CODE REVIEW

**Comment by boxcla** [comment]
Hi @anthonywee, thanks for the pull request. Before we can merge it, we need you to sign our Contributor License Agreement. You can do so electronically here: http://opensource.box.com/cla

Once you have signed, just add a comment to this pull request saying, "CLA signed". Thanks!

**Comment by anthonywee** [comment]
CLA Signed

**Comment by boxcla** [comment]
Verified that @anthonywee has just signed the CLA. Thanks, and we look forward to your contribution.

**Comment by potrebic** [comment]
Looks good to me. +1

File: flaky/_flaky_plugin.py [hunk 1]
```diff
@@ -233,7 +233,11 @@ def _add_flaky_report(self, stream):
             `file`
         """
         stream.write('===Flaky Test Report===\n\n')
-        stream.write(self._stream.getvalue())
+        value = self._stream.getvalue()
+        try:
+            stream.write(value)
+        except UnicodeEncodeError:
+            stream.write(value.encode('utf-8', 'replace'))
```
Comment (jmoldow) [inline flaky/_flaky_plugin.py]: Why not ``stream.write(ensure_unicode_string(self._stream.getvalue()))``? This seems to duplicate the behavior.

**Comment by jmoldow** [comment]
:+1: :+1: if you make the change to use ``ensure_unicode_string``, or explain why we shouldn't use it.

**Comment by Jeff-Meadows** [comment]
Agree with @jmoldow. Also, can you add a bugfix note under the upcoming section of the changelog?

**Comment by anthonywee** [comment]
I think we are looking at different cases: UnicodeDecodeError Vs. UnicodeEncodeError. ensure_unicode_string() tries to convert a string to unicode, and if that fails with a UnicodeDecodeError, it will *decode* the string with utf-8. In this situation, I believe there are cases where writing to the stream will try to convert it to a byte string, which may fail with a UnicodeEncodeError if it contains non-ascii characters. With the change, we will now *encode* the string into a byte string using utf-8 and execute the write again.

Sure, I can make a change to the HISTORY.rst file.

**Comment by anthonywee** [comment]
I've created a new pull request #49, and will close this one.

**Pull request closed without merging**